game/src/main/kotlin/content/skill/summoning/Crafting.kts

```diff
@@ -0,0 +1,321 @@
+ *
+ * @return: An [Item] with the charm id and amount needed to craft the given [pouch]
+ */
+fun getCharms(pouch: Item): Item {
+    val charmItemId: Int = pouch.def["summoning_charm_id"]
+    val charmAmount: Int = pouch.def["summoning_charm_amount"]
+    return Item(itemDefinitions.get(charmItemId).stringId, charmAmount)
+}
+
+/**
+ * Gets the pouch and amount needed to craft the given familiar pouch.
+ *
+ * @param pouch: The pouch being crafted.
+ *
+ * @return: An [Item] with the pouch id and amount needed to craft the given [pouch]
+ */
+fun getPouches(pouch: Item): Item {
+    val pouchItemId: Int = pouch.def["summoning_pouch_id"]
+    val pouchAmount: Int = pouch.def["summoning_pouch_amount"]
+    return Item(itemDefinitions.get(pouchItemId).stringId, pouchAmount)
+}
+
+/**
+ * Gets the maximum number of pouches the player can craft with the items they currently have in their
+ * inventory.
+ *
+ * @param player: The player
+ * @param shards: The shard [Item] with the associated amount of shards needed to craft the pouch
+ * @param charm: The charm [Item] with the associated amount of charms needed to craft the pouch
+ * @param pouch: The pouch [Item] with the associated amount of pouches needed to craft the pouch
+ * @param tertiaries: The list of 1 or more tertiary ingredient [Item]s needed for the pouch and the associated amount
+ * needed to craft the pouch.
+ *
+ * @return: The total number of pouches the player can craft
+ */
+fun maxCraftable(player: Player, shards: Item, charm: Item, pouch: Item, tertiaries: List<Item>): Int? {
+    if (tertiaries.isEmpty()) {
+        logger.info { "No tertiaries were provided." }
+        return null
+    }
+
+    val charmCount = player.inventory.count(charm.id).floorDiv(charm.amount)
+    val shardCount = player.inventory.count(shards.id).floorDiv(shards.amount)
+    val pouchCount = player.inventory.count(pouch.id).floorDiv(pouch.amount)
+    var minTertiaries = Int.MAX_VALUE
+    tertiaries.forEach { tertiary ->
+        val count = player.inventory.count(tertiary.id).floorDiv(tertiary.amount)
+        if (count < minTertiaries)
+            minTertiaries = count
+    }
+
+    return minOf(charmCount, shardCount, pouchCount, minTertiaries)
+}
+
+/**
+ * Sends a message to the given player with the ingredients needed for crafting a pouch.
+ *
+ * @param player: The player to send the message to
+ * @param enumIndex: The index of the clicked pouch in the "summoning_pouch_ids_1" enum.
+ */
+fun sendIngredientMessage(player: Player, enumIndex: Int) {
+    val realPouchId = enums.get("summoning_pouch_ids_1").getInt(enumIndex)
+    var ingredientString = enums.get("summoning_pouch_crafting_ingredient_strings").getString(realPouchId)
+
+    // The Rune Minotaur's ingredient string is incorrect in the cache.
+    if (realPouchId == itemDefinitions.get("rune_minotaur_pouch").id) {
+        ingredientString = "This pouch requires 1 runite bar, 1 blue charm and 100 spirit shards."
+    }
```

Comment:
Can't this now be fixed in FixItems ?

Reply:
Merged

Reply:
I've added `FixEnums` and I think it should work as-is, but it's giving an error.

```
Exception in thread "main" java.nio.BufferOverflowException
	at java.base/java.nio.Buffer.nextPutIndex(Buffer.java:725)
	at java.base/java.nio.HeapByteBuffer.put(HeapByteBuffer.java:216)
	at world.gregs.voidps.buffer.write.BufferWriter.writeByte(BufferWriter.kt:16)
	at world.gregs.voidps.buffer.write.Writer.writeString(Writer.kt:57)
	at world.gregs.voidps.cache.definition.encoder.EnumEncoder.encode(EnumEncoder.kt:44)
	at world.gregs.voidps.tools.cache.FixEnums.fix(FixEnums.kt:53)
	at world.gregs.voidps.tools.cache.FixEnums.main(FixEnums.kt:70)
```

Reply:
Needs a bigger BufferWriter size